docs: Describe server feature set (#1774)
CommunitySolidServer/CommunitySolidServer: documentation/markdown/features.md

# Features

Below is a non-exhaustive listing of the features available to a server instance,
depending on the chosen configuration.

> The core feature of the CSS is how it uses Components.js to configure its components,

Reviewer: ```suggestion
The core feature of the CSS is how it uses dependency injection to configure its components,
```

(and then later the framework)

so any of the features below can always be adapted or replaced with custom components if required.
See this [tutorial](https://github.com/CommunitySolidServer/tutorials/blob/main/custom-configurations.md)
and/or this [example repository](https://github.com/CommunitySolidServer/hello-world-component)
for more information on that.

To generate configurations with some of these features enabled/disabled,
you can use the [configuration generator](https://communitysolidserver.github.io/configuration-generator/).

## Authentication

Clients are identified based on the contents of DPoP tokens,
as described in the [Solid-OIDC specification](https://solidproject.org/TR/oidc).

> Alternatively, for debugging purposes the server can be configured to always identify the client as a fixed WebID,

Reviewer: The way it is currently written, sounds very limiting. I would use a couple of more words to describe that the DI framework also allows configuring dummy components for debugging, development, or experimentation purposes. One of those options is a fixed WebID, others are…

or to allow the WebID to be set directly in the `Authorization` header.
These can be configured by changing the `ldp/authentication` import in your configuration.

## Authorization

> Two different authorization are supported for determining who has access to resources:

Reviewer: ```suggestion
Two authorization mechanisms are implemented for determining who has access to resources:
```

* [Web Access Control](https://solidproject.org/TR/wac)
* [Access Control Policy](https://solidproject.org/TR/acp)

Or the server can be configured to not have any kind of authorization and allow full access to all resources.

> ## Solid Protocol

Reviewer: This section made me wonder whether we perhaps want bullets for all of it. Perhaps try and see if it is more scannable (possibly with some words in bold).

Author: Would it be more readable if everything was in bullets? I would think it then doesn't have that much use. Bolding keywords might help indeed.

Reviewer: Just suggestions, whatever you think works best.

> The [Solid Protocol](https://solidproject.org/TR/protocol) is supported, although it can be disabled if requested.

Reviewer: I don't understand this sentence; what would happened if the Protocol was disabled? It's not just the LDPish interface, but also auth.

Author: Currently disabling one would also disable the other as there is only the LDP interface to access authorized data.

Reviewer: Perhaps best to not mention that then?

> Requests to the server support content negotiation for most of the commonly used RDF formats.

Reviewer: ```suggestion
Requests to the server support content negotiation for common RDF formats.
```

Binary [range headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) are supported.

[ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) and Last-Modified headers are supported.
These can be used for [conditional requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests).

`PATCH` requests targeting RDF resources can be made with N3 Patch or SPARQL Update bodies.

The server can be configured to store data in memory, on the file system, or through a SPARQL endpoint.
Similarly, the locking system that is used to prevent data conflicts
can be configured to store locks in memory, on the file system, or in a Redis store, or it can be disabled.

Multiple worker threads can be used when starting the server.

## Account management

Accounts can be created on the server with which users can perform the following actions,

> either through an API or HTML pages:

Reviewer: ```suggestion
either through a JSON or HTML API:
```

* Add email/password combinations which can be used to log in.
* Create pods, which are containers on the server over which the user has full control.
* Link WebIDs to the account. When using [Solid-OIDC](https://solidproject.org/TR/oidc),

>   the user can use any of these to identify as.

Reviewer 2: ```suggestion
  the user can identify as any of these.
```

>   For external WebIDs the server requires the user to add a triple as identification,

Reviewer 2: ```suggestion
  For external WebIDs, the server requires the user to add a triple as identification,
```

  but this can be disabled if needed.
* Create client credentials, which can be used to authenticate without using the browser.
  More information on these can be found [here](../usage/client-credentials.md).
* It is possible to use Solid-OIDC to limit access to certain parts of the account API.
  More information on this can be found [here](../usage/identity-provider.md#access).

Using these accounts, the server can generate tokens
to support [Solid-OIDC](https://solidproject.org/TR/oidc) authentication.

### Pods

The server keeps track of the pod owners,
which is a list of WebIDs which have full control access over all resources contained within.

> Owners can be added and removed from a pod

Reviewer 2: ```suggestion
Owners can be added to and removed from a pod.
```

> Pod URLs can be minted in two different formats:

Reviewer 2: ```suggestion
Pod URLs can be minted as either
```
